File: src/main/java/ro/sopa/statistifier/service/TrackListenService.java
package ro.sopa.statistifier.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import ro.sopa.statistifier.db.repository.TrackListenRepository;
import ro.sopa.statistifier.service.dto.ArtistListens;
import ro.sopa.statistifier.service.model.ArtistTimelineEntry;
import ro.sopa.statistifier.service.model.ArtistTimelinePoint;

import java.time.Instant;
import java.time.Period;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.TemporalField;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class TrackListenService {

    @Autowired
    private TrackListenRepository trackListenRepository;

    @Value("${dbType}")
    private String dbType;

    private static DateTimeFormatter formatter = DateTimeFormatter.ISO_DATE;

    public List<ArtistListens> findTopArtists(String username, int number) {
        List<Object[]> artistsAndCounts = trackListenRepository.findTopArtists(username, number, dbType);

        return artistsAndCounts.stream().map(this::toListens).collect(Collectors.toList());
    }

    public List<ArtistTimelineEntry> findArtistsTimeline(String username, ZonedDateTime from, ZonedDateTime to, int nrSteps) {

        if(to == null) {
            to = ZonedDateTime.now();
        }

        if(from == null) {
            from = trackListenRepository.findOldestListenDate(username);
        }

        // așe pe scurt
        List<String> artists = trackListenRepository.findTopArtists(username, 20, dbType).stream().map(o -> (String) o[0]).collect(Collectors.toList());

        ZoneId zoneId = from.getZone();

        long fromMilis = from.toInstant().getEpochSecond();
        long toMilis = to.toInstant().getEpochSecond();

        long interval = (toMilis - fromMilis) / nrSteps;

        List<ArtistTimelineEntry> entries = new ArrayList<>();

        for (String artist : artists) {

            ArtistTimelineEntry entry = new ArtistTimelineEntry(artist);
            entries.add(entry);

            List<ZonedDateTime> dates = trackListenRepository.findListenDates(username, artist, from, to);

            // divide the times based on nr of steps
            for (long currentMilis = fromMilis; currentMilis < toMilis; currentMilis += interval) {
                ZonedDateTime min = ZonedDateTime.ofInstant(Instant.ofEpochSecond(currentMilis), zoneId);
                ZonedDateTime max = ZonedDateTime.ofInstant(Instant.ofEpochSecond(currentMilis + interval), zoneId);
                String stepDate = min.format(formatter);
                long count = dates.stream().filter(d -> d.isAfter(min) && d.isBefore(max)).count();
                entry.add(new ArtistTimelinePoint(stepDate, count));
            }
        }

        return entries;

    }

    private ArtistListens toListens(Object[] objects) {
        return new ArtistListens((String) objects[0], ((Number) objects[1]).intValue());
    }

}

File: src/main/kotlin/ro/sopa/lastfm/service/TrackListenService.kt
package ro.sopa.lastfm.service

import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import ro.sopa.lastfm.db.ArtistCountDto
import ro.sopa.lastfm.db.repository.TrackListenRepository
import ro.sopa.lastfm.db.repository.TrackListenRepositoryExtensions.findTopArtists
import ro.sopa.lastfm.service.dto.ArtistListens
import ro.sopa.lastfm.service.dto.ArtistTimelineEntry
import ro.sopa.lastfm.service.dto.ArtistTimelinePoint
import java.time.Instant
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

@Service
class TrackListenService(val trackListenRepository: TrackListenRepository, val artistService: ArtistService) {

    @Value("\${dbType}")
    private lateinit var dbType: String

    private val formatter = DateTimeFormatter.ofPattern("dd MMM yyyy")

    fun findTopArtists(userId: Int, number: Int): List<ArtistListens> {
        val artistsAndCounts: List<ArtistCountDto> = trackListenRepository.findTopArtists(userId, number, dbType)
        return artistsAndCounts.map { dto -> toListens(dto) }
    }

    fun findArtistsTimeline(
        userId: Int,
        from: ZonedDateTime?,
        to: ZonedDateTime?,
        nrSteps: Int,
        nrArtists: Int
    ): List<ArtistTimelineEntry> {

        var fromFinal = from ?: trackListenRepository.findOldestListenDate(userId) ?: ZonedDateTime.now()
        var toFinal: ZonedDateTime = to ?: ZonedDateTime.now()

        val artists: List<ArtistCountDto> = trackListenRepository.findTopArtists(userId, nrArtists, dbType, fromFinal, toFinal)

        val zoneId = fromFinal.zone
        val fromMilis = fromFinal.toInstant().epochSecond
        val toMilis = toFinal.toInstant().epochSecond
        val interval = (toMilis - fromMilis) / nrSteps

        val entries: MutableList<ArtistTimelineEntry> = mutableListOf()

        var artistMap = artistService.findIdToNameMap(artists.map { a -> a.getArtistId() })

        for (artist in artists) {
            val entry = ArtistTimelineEntry(artistMap[artist.getArtistId()]!!)
            val dates = trackListenRepository.findListenDates(userId, artist.getArtistId(), fromFinal, toFinal)

            // divide the times based on nr of steps
            var currentMilis = fromMilis
            while (currentMilis < toMilis) {
                val min = ZonedDateTime.ofInstant(Instant.ofEpochSecond(currentMilis), zoneId)
                val max = ZonedDateTime.ofInstant(Instant.ofEpochSecond(currentMilis + interval), zoneId)
                val stepDate = min.format(formatter)

                val count = dates.filter { d -> d.isAfter(min) && d.isBefore(max) }.count().toLong()

                entry.points.add(ArtistTimelinePoint(stepDate, count))
                currentMilis += interval
            }
            entries.add(entry)
        }
        return entries
    }

    private fun toListens(dto: ArtistCountDto): ArtistListens = ArtistListens(dto.getArtistId(), dto.getCount())

}